routing/missioncontrol_store.go

```diff
@@ -0,0 +1,232 @@
+	return b.db.Update(func(tx *bbolt.Tx) error {
+		if err := tx.DeleteBucket(resultsKey); err != nil {
+			return err
+		}
+
+		_, err := tx.CreateBucket(resultsKey)
+		return err
+	})
+}
+
+// fetch returns all results currently stored in the database.
+func (b *missionControlStore) fetch() ([]*paymentResult, error) {
+	var results []*paymentResult
+
+	err := b.db.View(func(tx *bbolt.Tx) error {
+		resultBucket := tx.Bucket(resultsKey)
+		results = make([]*paymentResult, resultBucket.Stats().KeyN)
+
+		i := 0
+		return resultBucket.ForEach(func(k, v []byte) error {
+			result, err := deserializeResult(k, v)
+			if err != nil {
+				return err
+			}
+
+			results[i] = result
+			i++
+
+			return nil
+		})
+
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("Fetched from db: %v results", len(results))
+
+	return results, nil
+}
+
+// deserializeResult deserializes a payment result.
+func deserializeResult(k, v []byte) (
+	*paymentResult, error) {
+
+	// Parse payment id.
+	result := paymentResult{
+		id: byteOrder.Uint64(k[8:]),
+	}
+
+	r := bytes.NewReader(v)
+
+	// Read timestamps, success status and failure source index.
+	var (
+		timeFwd, timeReply uint64
+		dbFailureSourceIdx int32
+	)
+
+	err := channeldb.ReadElements(
+		r, &timeFwd, &timeReply, &result.success, &dbFailureSourceIdx,
+	)
+	if err != nil {
+		return nil, err
+	}
+	result.timeFwd = time.Unix(0, int64(timeFwd)).UTC()
+	result.timeReply = time.Unix(0, int64(timeReply)).UTC()
+	if dbFailureSourceIdx != -1 {
+		failureSourceIdx := int(dbFailureSourceIdx)
+		result.failureSourceIdx = &failureSourceIdx
+	}
+
+	// Read route.
+	route, err := channeldb.DeserializeRoute(r)
+	if err != nil {
+		return nil, err
+	}
+	result.route = &route
+
+	// Read failure.
+	failureFlag, err := r.ReadByte()
+	if err != nil {
+		return nil, err
+	}
+
+	if failureFlag != 0 {
+		result.failure, err = lnwire.DecodeFailure(r, 0)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &result, nil
+}
+
+// AddResult adds a new result to the db.
+func (b *missionControlStore) AddResult(rp *paymentResult) error {
+	return b.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(resultsKey)
+
+		// Prune oldest entries.
+		if b.maxRecords > 0 {
+			count := bucket.Stats().KeyN
+			for count >= b.maxRecords {
+				cursor := bucket.Cursor()
+				cursor.First()
+				cursor.Delete()
+
+				count--
+			}
+		}
+
+		// Write timestamps, success status and failure source index.
+		var b bytes.Buffer
+
+		var dbFailureSourceIdx int32
+		if rp.failureSourceIdx == nil {
+			dbFailureSourceIdx = -1
+		} else {
+			dbFailureSourceIdx = int32(*rp.failureSourceIdx)
+		}
+
+		err := channeldb.WriteElements(
+			&b,
+			uint64(rp.timeFwd.UnixNano()),
+			uint64(rp.timeReply.UnixNano()),
+			rp.success, dbFailureSourceIdx,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Write route.
+		if err := channeldb.SerializeRoute(&b, *rp.route); err != nil {
+			return err
+		}
+
+		// Write failure.
```

Comment:
Possibly use `WriteVarBytes` here instead? The failure will just be blank upon read if nothing was specified.

Reply:
Converted to `WriteVarBytes`. I also discovered that I was serializing data that included length and padding. Created a separate commit that defines slimmer encode/decode functions in `lnwire`.